CORE-12165: merge key rotation feature work

dickon opened the pull request:
Take new API update with additional Avro, kafka and config for crypto key rotation.

Add new RestResource called KeyRotationRestResource which is automatically injected into the REST procesosr and allows crypto unmanaged wrapping key rotation to be started.

Implement unmanaged key rotation, for now in a relatively simple way where we take work requests on Kafka then process them to completion. This won't necessarily be the best way for the database at scale, and we have a design for spreading the work out over a nested Kafka topic, but let's get something in that is simple and works first.

Add alias to WrappingKeyInfo. It's not clear why this wasn't here before. WrappingKeyInfo is a data transfer object (DTO) and it seems reasonable for it to hide DB implementation detals such as the synthetic primary ID field. Alias is also part of a compound key with generation, and both alias and generation are user-visible features of wrapping keys; for instance a user could specify wrapping key alias when creating  a signing key pair.

Making WrappingKeyEntity key_material updatable since that's what happens during key rotation.

Add state manager arguments to the crypto worker, so that later the crypto worker can use the state manager for status notifications.

anien committed:
First draft of the CryptoRewrapBusProcessor

dickon committed:
do a rewrap

dickon committed:
test rewrapWrappingKey got called

dickon committed:
create subscription and processor for rewrapping

anien committed:
Fix typo in the topic var

anien committed:
Get resource for rewrap subscription

anien committed:
Fix compilation issues - implement missing members

anien committed:
Add comments

anien committed:
Add first version of the REST API calls for the key rotation and tue CryptoRekeyBusProcessor

anien committed:
Return empty list instead of events for now

anien committed:
Update avro schemas names to align with API changes

anien committed:
Add new lines

anien committed:
Update kafka topic names

anien committed:
Align with API changes for individual key rotation request

anien committed:
Rename rotation to keyRotation

anien committed:
Add parameters to REST API for key rotation

anien committed:
Add param in the kdoc

anien committed:
Add publishing Kafka message to rest impl and add initialise method for the config required for publisher

anien committed:
Find the keys that need to be rewrapped

anien committed:
Add start and end timestamp for key rotation response, return response to a response topic

anien committed:
Rename cryptoConfig to messagingConfig as that is what we use

anien committed:
Update timestamps according changes in api repo - status now contains start timestamp

anien committed:
Limit the number of key rotations if requested by user

anien committed:
Add key rotation rest resource test plus do changes to key rotatino response

anien committed:
Add oldKeyAlias and newKeyAlias validation to the key rotation rest resource, update tests

anien committed:
Simplify initialisation step in key rotation rest resource

anien committed:
Update kdocs - add throws, update key rotation response with correct values

anien committed:
Add comment

anien committed:
Add first CryptoRekeyBusProcessor tests

anien committed:
Simplify tests by creating a getRecord function

anien committed:
Add comment in the wrapping repo test reagarding the alias

anien committed:
Add test for zero limit in key rotation request

anien committed:
Fix the way we retrieve tenantId, close wrapping repo, close publisher

anien committed:
Add first draft of the test that checks which tenant's wrapping repo contains key for rotation

anien committed:
Create functions for duplicate code in a test

anien committed:
Remove crypto config from the test as it is not needed

anien committed:
Update topic name for key rotation

dickon committed:
bump API version to alpha version

dickon committed:
remove check code since crypto is not available

dickon committed:
remove unused imports (to fix detekt)

dickon committed:
remove test cases that cover client side config checks we can't do

dickon committed:
remove unused imports

dickon committed:
hack: set up second root key

dickon committed:
start the new internal publisher only once

dickon committed:
eliminate !! and race; set up new publisher in a more consistent way

dickon committed:
split line in publisher lookup

dickon committed:
fix vnode holding ID generationi

dickon committed:
put alias into wrapping key info

For now doing rekeying directly in CryptoRekeyBusProcessor since I think the invidiual rekey messages need aliases in them

dickon committed:
remove updatedable=false from a few wrapping key columns

dickon committed:
extra debug code around wrapping keys, probably best to drop

dickon committed:
fix detekt

dickon committed:
add new wrapping repostiory fields and methods

dickon committed:
fix detekt

dickon committed:
fix typo

anien committed:
Fix tests in rekey bus processor after changes in the code

anien committed:
Fix the last rekey bus processor test - fix the entity manager mock for a new function

anien committed:
Apply changes of adding alias in wrapping key info elsewhere

dickon committed:
fix alias handling

dickon committed:
bump API

dickon committed:
fix alias handling

dickon committed:
remove debug code that did not have mocks

dickon committed:
stop doing check that needs new mock

dickon committed:
fix test

dickon committed:
add state manager to crypto service, so far not used for anything

dickon committed:
undo whitespace change

dickon committed:
undo typo

dickon committed:
be more graceful in a couple of ways when using AssertWithRetryBuilder that currently throws null pointer exception.

If the command is missed out, there's no useful default behaviour, so throw IllegalArgumentException with an explanation.

If the condition is missed out, simply check for 2xx status, which is a reasonable default

dickon committed:
CORE-15936: add logic to call unmanaged wrapping key rotaiton endpoint in e2e tests

anien committed:
Removed unneeded imports

dickon committed:
CORE-12165: revert state manager, finish support for multiple root keys in the config

driessamyn commented on components/crypto/crypto-rest/src/main/kotlin/net/corda/crypto/rest/KeyRotationRestResource.kt:
```diff
@@ -0,0 +1,96 @@
+    name = "Key Rotation API",
+    description = "Contains operations related to rotation of the wrapping keys.",
+    path = "wrappingkey",
+    minVersion = RestApiVersion.C5_1
```
This should be `C5_2` now

dickon replied:
done in my PR

driessamyn commented on components/crypto/crypto-rest/src/main/kotlin/net/corda/crypto/rest/KeyRotationRestResource.kt:
```diff
@@ -0,0 +1,96 @@
+     *
+     * @param oldKeyAlias Alias of the key to be rotated.
+     * @param newKeyAlias Alias of the new key the [oldKeyAlias] key will be rotated with.
+     * @param simulate If true, only work out the number of keys to rotate.
```
I thought simulation was out of scope?

dickon replied:
Still an open question on the design PR - https://github.com/corda/platform-eng-design/pull/549#discussion_r1272470205 

I don't know whether it's best to continue that discussion there or in the code, but this morning I replied on the design PR thread.

driessamyn replied:
Regardless of the outcome, I think we should remove it from the API until it's implemented.
If I understand it correctly, the main benefit would be the generate a "report" of which keys are impacted, and I feel there are better ways to do this than a "simulation" option, which still would generate a bunch of msg bus traffic, for example.

dickon replied:
Okay, doing that in my branch dickon/CORE-12165-key-rotation-no-limit

dickon replied:
done in feature branch

driessamyn commented on components/crypto/crypto-rest/src/main/kotlin/net/corda/crypto/rest/KeyRotationRestResource.kt:
```diff
@@ -0,0 +1,96 @@
+     * @param oldKeyAlias Alias of the key to be rotated.
+     * @param newKeyAlias Alias of the new key the [oldKeyAlias] key will be rotated with.
+     * @param simulate If true, only work out the number of keys to rotate.
+     * @param timeToLive The maximum number of milliseconds which should pass, after which the key rotation loop will exit.
+     * @param limit The maximum number of key rotations to do, after which the key rotation loop will exit.
```
Do we need these given the design is for each rotation to be handled separately?

dickon replied:
I feel we need some way to control the amount of work done during key rotation, and limit is the easy way to do it from our code, since the hassle is with the operator.  If cancellation gets implemented before this code ships I'm open to taken this out again, though both a limit feature and a way to cancel could be useful to customers.

dickon replied:
I'm taking out limit in a PR to the feature branch https://github.com/corda/corda-runtime-os/pull/5129/files

driessamyn commented on components/crypto/crypto-rest/src/main/kotlin/net/corda/crypto/rest/impl/KeyRotationRestResourceImpl.kt:
```diff
@@ -0,0 +1,157 @@
+        oldKeyAlias: String,
+        newKeyAlias: String,
+        simulate: Boolean,
+        timeToLive: Int,
+        limit: Int
```
These are said to be optional in the docs, but don't have a default value.
(But see comment above about whether we actually need them).

dickon replied:
Yes, I'd missed that.

dickon replied:
I'm removing these arguments in https://github.com/corda/corda-runtime-os/pull/5129/files

dickon replied:
done in feature branch

Commit:
Merge branch 'release/os/5.2' into feature/CORE-12165/key-rotation

driessamyn commented on components/crypto/crypto-rest/src/main/kotlin/net/corda/crypto/rest/impl/KeyRotationRestResourceImpl.kt:
```diff
@@ -0,0 +1,157 @@
+        publisher?.publish(listOf(Record(uploadTopic, requestId, keyRotationRequest)))
+            ?: throw ServiceUnavailableException("Key rotation resource has not been initialised.")
+
+        return ResponseEntity.accepted(KeyRotationResponse(requestId, oldKeyAlias, newKeyAlias))
```
Doesn't this need to wait for the future to complete (and handle a potential exception)?

dickon replied:
I don't think I understand the REST worker design well enough to say so I'd like to talk to @anien about this when she's in.  I think the idea was to crib what the other resources were doing.

anien replied:
This was implemented similarly to the start flow request. The main reason for having it like this is that we don't know how long the key rotation will take. So the idea was that we tell the user we have accepted the request and the key rotation status is the way to tell user how we are doing/what's the progress of the key rotation.

dickon replied:
Okay I think this is fine and what I expected for now. The alternative is to make the HTTP request synchronous, so it get a response until the work is done, but given the crypto worker may not be available and we are dispatching to Kafka I don't think that's wanted.

driessamyn replied:
But doesn't the future just wait for the publish to complete?
I could be wrong, but I think the way it currently is means that if the publish would fail, you would still get an ok response.

anien replied:
If the publisher is not initialised, it throws the exception. There is also a test case for it. I am not that amazingly familiar with the rest resources in Corda, so happy to discuss this if you think it should be done differently.

driessamyn commented on components/crypto/crypto-rest/src/test/kotlin/net/corda/crypto/rest/impl/KeyRotationRestResourceTest.kt:
```diff
@@ -0,0 +1,177 @@
+        whenever(publisherFactory.createPublisher(any(), any())).thenReturn(publisher)
+    }
+
+//    @Test
+//    fun `get key rotation status`() {
+//        TODO("Not yet implemented")
+//    }
+//
+//    @Test
+//    fun `get key rotation status for unknown requestID throws`() {
+//        TODO("Not yet implemented")
+//    }
```
I am assuming these will be filled in as part of the status work?

Instead it may be better to add them in and `@Disable` them with a reference to the JIRA.
Disabled tests are a bit more visible than ones commented out.

dickon replied:
will use @Disabled in my PR.

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:
```diff
@@ -0,0 +1,65 @@
+
+    @Suppress("NestedBlockDepth")
+    override fun onNext(events: List<Record<String, KeyRotationRequest>>): List<Record<*, *>> {
+        logger.info("received ${events.size} key rotation requests")
```
This should be debug level logging, not `info`.

dickon replied:
done on my https://github.com/corda/corda-runtime-os/pull/5134 (and probably the first one too)_

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:
```diff
@@ -0,0 +1,65 @@
+    override fun onNext(events: List<Record<String, KeyRotationRequest>>): List<Record<*, *>> {
+        logger.info("received ${events.size} key rotation requests")
+        events.map { it.value}.filterNotNull().forEach { request ->
+            logger.info("processing $request")
```
As above, change to `debug` level logging.

dickon replied:
done on https://github.com/corda/corda-runtime-os/pull/5134

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:
```diff
@@ -0,0 +1,65 @@
+            // tenantId in the request is useful ONLY for re-wrapping managed keys!
+            // For unmanaged (root) key we need to go through all vNodes and check if the oldKeyAlias is used there
+            // if yes, then we need to issue a new record for this key to be re-wrapped
+            // we get the tenantId from the vNode info (I hope).
```
Not necessarily because tenants are not always VNodes (they can be "P2P" for example, IIRC).

dickon replied:
That's correct, and I agree the comment should change.

dickon replied:
Comment fixed on https://github.com/corda/corda-runtime-os/pull/5134

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:
```diff
@@ -0,0 +1,65 @@
+            val virtualNodeInfo = virtualNodeInfoReadService.getAll() // Get all the virtual nodes
+            val virtualNodeTenantIds = virtualNodeInfo.map { it.holdingIdentity.shortHash.toString() }
+
+            val allTenantIds = virtualNodeTenantIds+listOf(CryptoTenants.CRYPTO, CryptoTenants.P2P, CryptoTenants.REST)
```
Should we have a central list of "all" non-vnode tenants?
Or use the `allClusterTenants` function?

dickon replied:
It turns out this was wrong, since the non-vnode tenants share a cluster database, so we only need to look at the table once, otherwise we end up rewrapping keys 3 times. So this code is now obsolete and the new version in https://github.com/corda/corda-runtime-os/pull/5134 has a comment on this.,

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:
```diff
@@ -0,0 +1,65 @@
+            val virtualNodeTenantIds = virtualNodeInfo.map { it.holdingIdentity.shortHash.toString() }
+
+            val allTenantIds = virtualNodeTenantIds+listOf(CryptoTenants.CRYPTO, CryptoTenants.P2P, CryptoTenants.REST)
+            logger.info("Found ${allTenantIds.size} tenants; first few are: ${allTenantIds.take(10)}")
```
Remove or change to debug level.

dickon replied:
done in https://github.com/corda/corda-runtime-os/pull/5134

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:
```diff
@@ -0,0 +1,65 @@
+                }
+            }.flatten()
+
+            val truncatedWrappingKeys = targetWrappingKeys.take(request.limit ?: Int.MAX_VALUE)
```
Why is the limit useful here?
I feel this is over-engineered and will cause end-user confusion, and call YAGNI on it.

dickon replied:
removed, even though I think it was useful, it's an architectural call.

dickon replied:
removed in https://github.com/corda/corda-runtime-os/pull/5129

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRewrapBusProcessor.kt:
```diff
@@ -0,0 +1,52 @@
+                )
+            records.add(Record(REWRAP_MESSAGE_RESPONSE_TOPIC, request.requestId, value))
+        }
+        return records
```
Is this to be replaced with the State Manager?

dickon replied:
It's a response that acks the rewrap request, not the status of the request. I was thinking we'd make lots of incremental updates to state manager as the work proceeded.

anien replied:
Yes, this was originally planned it might be used for the status, but without a clear idea. It's not needed at the moment and once status is implemented, state manager will be used.

driessamyn commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/WrappingRepository.kt:
```diff
@@ -19,7 +19,7 @@ interface WrappingRepository : Closeable {
      * @return The wrapping key that was persisted.
      */
-    fun saveKey(alias: String, key: WrappingKeyInfo): WrappingKeyInfo
+    fun saveKey(key: WrappingKeyInfo): WrappingKeyInfo
```
Remove alias from the KDoc

dickon replied:
done in my PR

dickon replied:
in feature branch

driessamyn commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/WrappingRepository.kt:
```diff
@@ -30,7 +30,7 @@ interface WrappingRepository : Closeable {
      * @return The wrapping key that was persisted.
      */
-    fun saveKeyWithId(alias: String, key: WrappingKeyInfo, id: UUID?): WrappingKeyInfo
+    fun saveKeyWithId(key: WrappingKeyInfo, id: UUID?): WrappingKeyInfo
```
Remove alias from the KDoc

dickon replied:
done in my pR

dickon replied:
in feature branch

driessamyn commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoService.kt:
```diff
@@ -579,15 +580,19 @@ open class SoftCryptoService(
             val oldParentKey = checkNotNull(unmanagedWrappingKeys.get(wrappingKeyInfo.parentKeyAlias)) {
                 "Unable to find parent key ${wrappingKeyInfo.parentKeyAlias} in the configured unmanaged wrapping keys"
             }
+            val newGeneration = wrappingKeyInfo.generation + 1
```
Is a race condition possible here in case `wrappingKeyInfo` is out of date?
Do we need optimistic locking?

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRekeyBusProcessor.kt:
```diff
@@ -0,0 +1,65 @@
+        private val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
+    }
+
+    override val keyClass: Class<String> = String::class.java
```
If I read it correctly, I think the key here is tenantId.
Does this make sense?
The implication is that all requests for a particular tenant ID will be going to the same Kafka partition (and hence the same DB processor).
An advantage could be batching (i.e. less DB connections), but I don't think we do that on the consumer side, and if we wanted to do batching we could do this here.
The disadvantage is that I don't think the tenantID is "random" enough , so if you have a single VNode with 1000s of keys, only 1 processor can process them.

I think it may make more sense to use a different field that is more random.

dickon replied:
No. The code is a bit confusing and I've improved it on my second PR https://github.com/corda/corda-runtime-os/pull/5134

On that branch, things are how I think they should be. The operator calls the REST worker, that creates one `KeyRotationRequest` which may be for the whole cluster or a single vnode. That's then picked up by a crypto worker, and that published a number of `IndvidiaulKeyRotationRequest` messages, potentially a large number. If lots of wrappings are found in the same tenant there will be many messages. As far as I know the work is then picked up by all the separate workers.

dickon replied:
5134 is merged, callng this resolved.

driessamyn reviewed (commented):
WIP review - given pair programming was used and all code already seen by 2 people, I only focused on high level review.
Mostly high level comments and questions.
Couple of questions about scope (requirement to support limit and "simulation" mode).

driessamyn commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/WrappingRepositoryImpl.kt:
```diff
@@ -39,14 +39,14 @@ class WrappingRepositoryImpl(
                         parentKeyReference = key.parentKeyAlias,
                     )
-                )
+                ).also { wrappingKeyEntity -> check(wrappingKeyEntity.generation == key.generation) }
```
Isn't this something that should be done in the DB rather than in code?
There is a standard pattern for optimistic locking which we use elsewhere and which is thread safe (this isn't).

dickon replied:
Yeah, I hope I've now dealt with the race in SoftCryptoService.createWrappingKey, and the other path that reaches here, in https://github.com/corda/corda-runtime-os/pull/5134

driessamyn commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/WrappingRepositoryImpl.kt:
```diff
@@ -39,14 +39,14 @@ class WrappingRepositoryImpl(
+                ).also { wrappingKeyEntity -> check(wrappingKeyEntity.generation == key.generation) }
             }.toDto().also {
-                logger.info("Storing wrapping key with alias $alias in tenant $tenantId")
+                logger.info("Storing wrapping key with alias ${key.alias} in tenant $tenantId")
```
If we are logging, is it worth adding the generation id as well?

dickon replied:
done in https://github.com/corda/corda-runtime-os/pull/5134

driessamyn commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/WrappingRepositoryImpl.kt:
```diff
@@ -58,6 +58,22 @@ class WrappingRepositoryImpl(
             }
         }
+
+    // TODO add test coverage
```
Is there a JIRA for this?
Any reason this wasn't included in the development from the start?

dickon replied:
test coverage added in https://github.com/corda/corda-runtime-os/pull/5134

driessamyn commented on components/crypto/crypto-softhsm-impl/src/main/kotlin/net/corda/crypto/softhsm/impl/WrappingRepositoryImpl.kt:
```diff
@@ -58,6 +58,22 @@ class WrappingRepositoryImpl(
+            it.createQuery(
+                "FROM ${WrappingKeyEntity::class.simpleName} AS k WHERE k.parentKeyReference = :alias",
+                WrappingKeyEntity::class.java
+            ).setParameter("alias", alias).resultStream.map {dao -> dao.toDto() }.toList().asSequence()
```
Why do we go from stream -> list -> stream again?

dickon replied:
no good reason, in https://github.com/corda/corda-runtime-os/pull/5134 it is stream -> sequence, since we usually pass around Kotlin types

driessamyn commented on components/crypto/crypto-softhsm-impl/src/test/kotlin/net/corda/crypto/softhsm/impl/SoftCryptoServiceCachingTests.kt:
```diff
@@ -208,6 +208,14 @@ class SoftCryptoServiceCachingTests {
 
             override fun findKeyAndId(alias: String): Pair<UUID, WrappingKeyInfo>? = TODO("Not needed")
 
+            override fun findKeysWrappedByAlias(alias: String): Sequence<WrappingKeyInfo> {
+                TODO("Not yet implemented")
+            }
+
+            override fun getKeyById(id: UUID): WrappingKeyInfo? {
+                TODO("Not yet implemented")
+            }
+
```
Is this intentionally not yet implemented?

dickon replied:
yes, this is test fixture and there's no test cases which use hit these endpoints, but the Kotlin compiler requires all overrides to be implemented.

driessamyn commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/application/crypto/SigningServiceImpl.kt:
```diff
@@ -120,6 +120,22 @@ class SigningServiceImpl @Activate constructor(
         return recordSuspendable({ cryptoFlowTimer("findMySigningKeys") }, operation)
     }
 
+    override fun signWithMyKeys(
+        bytes: ByteArray,
+        keys: MutableSet<PublicKey>,
+        signatureSpec: SignatureSpec
+    ): MutableSet<DigitalSignature.WithKeyId> {
+        TODO("Not yet implemented")
+    }
+
+    override fun signWithMyCategoryKeys(
+        bytes: ByteArray,
+        category: String,
+        signatureSpec: SignatureSpec
+    ): MutableSet<DigitalSignature.WithKeyId> {
+        TODO("Not yet implemented")
+    }
+
```
Is this related to key rotation?

dickon replied:
No, as I was saying on Slack, I think it's a git artifact of a revert since we last rebased, and I'm working on a personal branch to fix that. Hopefully github will get the merge right when that goes in.

dickon replied:
Removed by fixing what I think was a mistake in the merge Dries did from 5.2

driessamyn commented on gradle.properties:
```diff
@@ -45,8 +45,8 @@ commonsLangVersion = 3.12.0
-# Change to 5.2.0.xx-SNAPSHOT to pick up maven local published copy
-cordaApiVersion=5.2.0.2-beta+
+# Change to 5.1.0.xx-SNAPSHOT to pick up maven local published copy
+cordaApiVersion=5.1.0.38-alpha-1699268962803
```
FYI we should replace this with a 5.2 reference.

dickon replied:
done in https://github.com/corda/corda-api/pull/1349

dickon replied:
done

driessamyn commented on libs/application/application-db-setup/src/main/kotlin/net/corda/application/dbsetup/ConfigEntityFactory.kt:
```diff
@@ -42,8 +43,16 @@ class ConfigEntityFactory(
+                    smartConfigFactory.makeSecret(random.randomString(), "corda-master-wrapping-key-passphrase").root(),
+                    smartConfigFactory.makeSecret(random.randomString(), "corda-master-wrapping-key-salt").root()
+                ),
+                KeyDerivationParameters(
+                    smartConfigFactory.makeSecret(random.randomString(), "corda-master-wrapping-key-2-passphrase").root(),
+                    smartConfigFactory.makeSecret(random.randomString(), "corda-master-wrapping-key-2-salt").root()
+                )
+            ),
```
Why do we generate 2 when bootstrapping?

dickon replied:
So we've got enough to do a key rotation in the e2e tests. Bit of a hack, which might justify a comment since it isn't obvious enough.

dickon replied:
reworked in https://github.com/corda/corda-runtime-os/pull/5126/files so this comment is obsolete.

dickon replied:
done

anien replied:
We still have two keys here. Do we want to keep that?

driessamyn replied:
I think it might be quite confusing.

driessamyn reviewed (commented):
Some more minor comments.

dickon committed:
Merge remote-tracking branch 'origin/release/os/5.2' into feature/CORE-12165/key-rotation

# Conflicts:
#	gradle.properties

dickon committed:
use suitable API version

dickon committed:
update for API version

dickon committed:
Merge pull request #5138 from corda/dickon/CORE-12165-feature-branch-merge-5-2-2023-11-20

CORE-12165: merge latest 5.2 changes and update for latest key rotation API

dickon committed:
CORE-12165: key rotation no limit (#5129)

* bump API reference

* go to debug levecl

* change REST path to 5.2

* drop simulate, timeToLive and limit parameters

* remove limit and simulate flag

* remove typo

* use new API

* use new field name agreed in API change

* bump API version

dickon committed:
CORE-12165: restore nested kafka topics (#5134)

* bump API reference

* go to debug levecl

* remove limit and simulate flag

* remove typo

* restore nested topic

* prevent duplicated rotation of cluster wrapping keys

* simplify stream/sequence handling
add test case

* remove wildcard import

* remove test duplication

* fix formatting

* update for API changes

* fix rebase error

* fix formatting

dickon commented:
Matching PR for API change is https://github.com/corda/corda-api/pull/1348

mag1c-48 reviewed (dismissed):
LGTM - Build files only

charlieR3 reviewed (dismissed):
LGTM - network team files

anien committed:
Minor comment fixes, typos (#5141)

vkolomeyko commented on components/crypto/crypto-rest/src/main/kotlin/net/corda/crypto/rest/KeyRotationRestResource.kt:
```diff
@@ -0,0 +1,76 @@
+    path = "wrappingkey",
+    minVersion = RestApiVersion.C5_2
+)
+interface KeyRotationRestResource : RestResource {
```
Could you please add this new resource to `OpenApiCompatibilityTest` ?

It will also require Open API Baseline to be updated.

vkolomeyko commented on testing/e2e-test-utilities/src/main/kotlin/net/corda/e2etest/utilities/AssertWithRetryBuilder.kt:
```diff
@@ -10,9 +10,13 @@ class AssertWithRetryArgs {
     var interval: Duration = Duration.ofMillis(500)
     var startDelay: Duration = Duration.ofMillis(10)
     var command: (() -> SimpleResponse)? = null
-    var condition: ((SimpleResponse) -> Boolean)? = null
```
This seems like unrelated change. Would it not be better to do it as a separate PR?

dickon replied:
This merges a feature branch that has been open for months, and while I would normally like to keep PRs focussed on a single change I feel in this circumstance it's not worth the time to unpick changes.

vkolomeyko commented on components/crypto/crypto-rest/src/main/kotlin/net/corda/crypto/rest/KeyRotationRestResource.kt:
```diff
@@ -0,0 +1,76 @@
+     * of the API.
+     * @param config A config
+     */
+    fun initialise(config: Map<String, SmartConfig>)
```
Do we need this non-REST annotated method on the interface?

IMO better way of onboarding config is by subscribing to configuration change, please see: use of `configurationReadService.registerComponentForUpdates` in `VirtualNodeRestResourceImpl`.

anien replied:
The approach was taken from the `FlowRestResource`. I will take a look if we can simply use private function.

anien replied:
Fixed in [#5143](https://github.com/corda/corda-runtime-os/pull/5143)

vkolomeyko commented on components/crypto/crypto-rest/src/main/kotlin/net/corda/crypto/rest/KeyRotationRestResource.kt:
```diff
@@ -0,0 +1,76 @@
+package net.corda.crypto.rest
```
I think this new package requires OSGi export provisions or else this pluggable REST resource is not going to be discovered by REST Worker.

Also, it would probably make sense to do some Dev testing of the new REST endpoint using Combined Worker locally.

anien replied:
The new REST endpoints have already been tested locally in K8s cluster. The endpoints were available in swagger as well as called via curl command.
What do you mean that it might requires OSGi export please?

anien replied:
Fixed in #5143

driessamyn commented on components/crypto/crypto-service-impl/src/main/kotlin/net/corda/crypto/service/impl/bus/CryptoRewrapBusProcessor.kt:
```diff
@@ -0,0 +1,38 @@
+    override fun onNext(events: List<Record<String, IndividualKeyRotationRequest>>): List<Record<*, *>> {
+        return events.mapNotNull { it.value }.map { request ->
+            cryptoService.rewrapWrappingKey(request.tenantId, request.targetKeyAlias, request.newParentKeyAlias)
+            Record(
+                REWRAP_MESSAGE_RESPONSE_TOPIC,
+                request.requestId, IndividualKeyRotationResponse(
+                    request.requestId,
+                    request.tenantId,
+                    request.oldParentKeyAlias,
+                    request.newParentKeyAlias,
+                    request.targetKeyAlias,
+                    Instant.now()
+                )
+            )
```
Why do we need this response AVRO message and topic?
I'm guessing this will be replaced with the SM implementation, so wouldn't it make sense to take these away from the API change?

anien replied:
We agreed over slack discussion that after status using state manager is implemented, we will review all topics and avro schemas introduces in key rotation. Jira ticket is created for this.

dickon replied:
Removed for now.

anien committed:
CORE-12165: Update swagger baseline test, remove function from rest resource (#5143)

* Remove initialise function from the rest resource interface.

* Update swagger baseline test with new rest resource.

* Add package files for OSGi export.

anien committed:
CORE-12165: Use UUID as key to spread the work evenly between Kafka partitions (#5146)

* User random key for kafka topic to spread the work between partitions

* Use random key for kafka topic to spread the work between partitions on producer side too

dickon committed:
CORE-12165: merge 5.2 changes and use the expected new version of the feature branch on the API (#5151)

* CORE-18072 CORE-18263 fix resolveStateRef logic and call (#5071)


- findSignedLedgerTransaction explicitly called the resolveStateRefs function which resulted in opening two DB connections at the same time. This has been modified to a direct call to the repository instead.
- The resolveStateRefs function in the state query service uses a cache but we always fetch all of the state refs instead of just the non-cached ones.

* CORE-17972: implementation of send & receive tx (#5042)

implemented logic for sending and receiving a transaction and verification for the transaction received.

* CORE-17961: Allow Time & Multiple Metadata Filters (#5139)

- Allow filtering states by last updated timestamp and multiple
  metadata filters (both conjunctive and disjunctive operations).
- Update 'ScheduledTaskProcessor' to use the new feature instead of
  manually filtering states.

* CORE-17821 Removed Kafka implementation from CryptoProcessorImpl (#5135)

This PR removes any deprecated or duplicated code, particularly Kafka-based implementations, in the CryptoProcessor which was created during the migration from Kafka to RPC external event communications.

* CORE-18413: Increase timeout for vNode operations for E2E tests (#5144)

* tidy up merge

* Return a list from findKeysWrappedByAlias to avoid using the Stream after it is closed

* use new API version

* remove old imports

* fix merge

* fix

* fix tests

* fix detekt

* Revert "CORE-12165: Use UUID as key to spread the work evenly between Kafka partitions (#5146)"

This reverts commit 3490b404

* fix revert and tests

* CORE-12459: Configure Corda dedicated vNode admin fix (#5123)

Allows a user that is not the cluster user to sync the vNode on an external DB

* CORE-18152/CORE-17601: Remove Kryo Pool Upper Limit (#5145)

* CORE-18352 Fix app-simulator optional database parameters (#5105)

The app-simulator’s database parameters are optional when running in SENDER mode. This change fixes an issue where the Sender fails to start if these are not provided.

* fix test

* use final 5.2.0.5 API version

---------

Co-authored-by: nkovacsx <57627796+nkovacsx@users.noreply.github.com>
Co-authored-by: Jenny Yang <135036209+jennyang-r3@users.noreply.github.com>
Co-authored-by: Juan José Ramos <jujoramos@users.noreply.github.com>
Co-authored-by: Ben Millar <44114751+ben-millar@users.noreply.github.com>
Co-authored-by: Viktor Kolomeyko <viktor.kolomeyko@r3.com>
Co-authored-by: Ben Yip <ben.yip@r3.com>
Co-authored-by: Yash Nabar <yash.nabar@r3.com>

dickon committed:
Merge remote-tracking branch 'origin/release/os/5.2' into dickon/CORE-12165-zero-merge

# Conflicts:
#	gradle.properties
#	processors/crypto-processor/src/main/kotlin/net/corda/processors/crypto/internal/CryptoProcessorImpl.kt

dickon committed:
Merge pull request #5155 from corda/dickon/CORE-12165-zero-merge

CORE-12165: record merge

vkolomeyko reviewed (approved):
LGTM (on behalf of @corda/rest)

emilybowe reviewed (approved):
LGTM (flow changes only)

driessamyn reviewed (approved):
LGTM

lankydan reviewed (approved):
Approved on the basis that no ledger code seems to have changed.

mag1c-48 reviewed (approved):
LGTM - build

wzur-r3 reviewed (approved):
LGTM for settings.gradle owned by BLT

charlieR3 reviewed (approved):
LGTM - network team files
